circuit.py

```diff
@@ -97,18 +93,71 @@ def apply_transformation(self) -> None:
-
-                # if isinstance(self.ops[index + 1], Measurement):
-                #     break
-                
-                self.commute_rotation(index)
+            while index + 1 < len(self):
+                self.commute_pi_over_four_rotation(index)
                 index += 1
+            self.ops.pop()
+        
+        # Remove all pi/4 rotations from the circuit
+        self.remove_y_operators_from_circuit()
+
+
+    def remove_y_operators_from_circuit(self) -> None:
+        """
+        Removes Y operators from pi/8 and measurement blocks. To be called after pi/4 rotations 
+        have been commuted to the end of the circuit.
+
+        """
+        i = 0 
 
+        while i < len(self.ops):
+            pauli_block = self.ops[i]
 
-    def commute_rotation(self, index: int) -> None:
+            if isinstance(pauli_block, Measurement) or (pauli_block.rotation_amount in {Fraction(1,8), Fraction(-1,8)}):
+                y_op_indices = list()
+                
+                # Find Y operators and modify them into X operators 
+                for j in range(self.qubit_num):
+                    if pauli_block.ops_list[j] == PauliOperator.Y:
+                       y_op_indices.append(j)
+                       pauli_block.ops_list[j] = PauliOperator.X
+
+                if y_op_indices:
+                    right_block_indices = list()
+
+                    # For even numbers of Y operators, add 2 additional pi/4 rotations (one on each side)
+                    if len(y_op_indices) % 2 == 0:
+                        first_operator = y_op_indices.pop(0)
+                        self.add_single_operator(first_operator, PauliOperator.Z, Fraction(1,4), i)
+                        self.add_single_operator(first_operator, PauliOperator.Z, Fraction(-1,4), i+2)
+                        right_block_indices.append(i+4)
+                        i += 1
+
+                    # Add 2 pi/4 rotations on each side of the Pauli block
+                    new_block = [PauliOperator.Z if i in y_op_indices else PauliOperator.I for i in range(self.qubit_num)]                    
+                    self.add_pauli_block(Rotation.from_list(new_block, Fraction(1,4)), i)
+                    self.add_pauli_block(Rotation.from_list(new_block, Fraction(-1,4)), i+2)
+                    right_block_indices.append(i+2)
+                    i += 1
+
+                    # Commute the right (new) pi/4 rotations towards the end of the circuit
+                    for right_block_index in right_block_indices:
+                        while right_block_index + 1 < len(self.ops):
+                            self.commute_pi_over_four_rotation(right_block_index)
+                            right_block_index += 1
+                        self.ops.pop()
+            else:
+                # This is assuming pi/4 rotations (not including ones from this operation) 
+                # have been commuted to the end of the circuit.
+                break
```

Comment:
Now that I think about it more carefully, would it be very challenging to make this transformation work without assuming that the Litinski Transform has already been applied? There might be reasons why we wouldn't want to always apply it, like if we are debugging other stuff and we still want operations to stay.

Reply:
Intuitively I would think that all we have to do is turn the break into a continue and increase the counter
```suggestion
                continue
```

Reply:
Of course this doesn't work for pi/4 rotations, but I guess that's ok for now we can deal with thatas it becomes necessary, since it's as simple of a change

Reply:
Originally `apply_transformation()` does not remove the `pi/4` rotations after they are moved to the end of the circuit, that is why I left the the `break` comment there just to terminate the while loop early. Implement removing the `pi/4` rotations was the last minute decision - I was conflicted on whether or not to remove information from the circuit (in case we want to reverse the operation).

Reply:
Anyway it was me overthinking stuffs - discussions adding options to reverse `apply_transformation()` can happen later. I will go ahead and merge the branch.